File: src/crossPageMetaDescription.ts
/**
 * crossPageMetaDescription.ts — AGGREGATES-LAYER detector. T76.
 *
 * Sister to `crossPageTitle.ts`: walks accumulated cross-page
 * state to surface defects no single-page snapshot can see.
 * Where crossPageTitle catches duplicate `<title>` across a
 * journey, this catches duplicate `<meta name="description">`.
 *
 * Real-world impact: Google explicitly filters near-duplicate
 * meta descriptions in search results — the affected pages are
 * effectively invisible. Social-share previews collapse into a
 * single card if every page advertises the same blurb. The
 * baseline SEO doctrine is one description per route.
 *
 * Findings:
 *
 *   - meta-description.cross-page-dup     warn
 *     Two or more pages in the journey share the same trimmed
 *     `<meta name="description">` content.
 *
 * Out of scope:
 *   * Pages with NO description — covered by per-page
 *     `meta-description.missing` finding.
 *   * Empty descriptions — covered by `meta-description.empty`.
 *
 * No Rust mirror in v1 (same rationale as crossPageTitle).
 *
 * Pattern reuse: this is the SECOND aggregates-layer detector.
 * The shape is intentionally identical to crossPageTitle so
 * future ones can copy this template — accumulator + record +
 * detect-once-after-loop. If a third aggregates detector lands,
 * extract a generic `dupGroupDetector(accumulator, kind, label)`
 * helper.
 */

export interface CrossPageMetaDescriptionFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface CrossPageMetaDescriptionAccumulator {
  /** Ordered list of {url, description} as the journey visits each page. */
  entries: Array<{ url: string; description: string }>;
}

/** Construct an empty accumulator. */
export function newCrossPageMetaDescriptionAccumulator():
  CrossPageMetaDescriptionAccumulator {
  return { entries: [] };
}

/**
 * Record a page's description. Caller passes the RAW description
 * (this function trims). Empty descriptions are skipped — the
 * per-page meta-description.empty / meta-description.missing
 * findings handle that case better than reporting "every empty
 * description is a duplicate".
 */
export function recordPageMetaDescription(
  acc: CrossPageMetaDescriptionAccumulator,
  url: string,
  description: string,
): void {
  const trimmed = description.trim();
  if (!trimmed) return;
  acc.entries.push({ url, description: trimmed });
}

/**
 * Walk the accumulator. For each description that appears on > 1
 * distinct URL, emit one warn finding listing the URLs that
 * share it.
 */
export function detectCrossPageMetaDescriptionDuplicates(
  acc: CrossPageMetaDescriptionAccumulator,
): CrossPageMetaDescriptionFinding[] {
  // T76 cycle 59: dedupe by URL pathname so journeys that revisit
  // the same path with different query strings (e.g. the
  // themes journey hitting `/?theme=dark`, `/?theme=light`, …)
  // don't get treated as distinct pages. See crossPageTitle.ts
  // for the matching fix.
  const pathKey = (u: string): string => {
    try {
      const parsed = new URL(u);
      return parsed.origin + parsed.pathname;
    } catch {
      return u;
    }
  };

  const groups = new Map<string, Map<string, string>>();
  for (const { url, description } of acc.entries) {
    let urls = groups.get(description);
    if (!urls) {
      urls = new Map<string, string>();
      groups.set(description, urls);
    }
    const k = pathKey(url);
    if (!urls.has(k)) urls.set(k, url);
  }

  const out: CrossPageMetaDescriptionFinding[] = [];
  for (const [description, urls] of groups) {
    if (urls.size < 2) continue;
    const urlList = Array.from(urls.values());
    // Truncate the description in the detail for readability —
    // these can be 150+ chars per the metaDescription floor.
    const preview = description.length > 80
      ? `${description.slice(0, 80)}…`
      : description;
    out.push({
      severity: 'warn',
      kind: 'meta-description.cross-page-dup',
      detail: `${urls.size} distinct URL(s) share the same <meta name="description"> content '${preview}'. Google filters near-duplicate descriptions in search results — affected pages become effectively invisible. Social-share preview cards collapse. Set a unique, page-specific description per route. URLs: ${urlList.slice(0, 5).join('; ')}`,
      evidence: { description, urlCount: urls.size, urls: urlList },
    });
  }

  return out;
}

File: crates/crawler-detectors/src/cross_page_meta_description.rs
//! `cross_page_meta_description` — aggregates-layer detector for
//! duplicate `<meta name="description">` content across pages.
//!
//! Mirror of `src/crossPageMetaDescription.ts`. Sister to
//! `cross_page_title`. One finding kind:
//!
//!   * `meta-description.cross-page-dup` warn
//!
//! Real-world impact: Google filters near-duplicate descriptions in
//! search results — affected pages become effectively invisible.
//! Social-share preview cards collapse.
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector, no I/O.

use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Ordered list of (url, description) records the crawler
/// accumulates as it walks a journey.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[non_exhaustive]
pub struct CrossPageMetaDescriptionAccumulator {
    /// One entry per page visit, in journey order. Empty
    /// descriptions are filtered at record time.
    pub entries: Vec<CrossPageMetaDescriptionEntry>,
}

/// One recorded (URL, description) pair.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[non_exhaustive]
pub struct CrossPageMetaDescriptionEntry {
    /// URL as visited (preserves query string for evidence).
    pub url: String,
    /// Trimmed meta-description content.
    pub description: String,
}

/// Construct an empty accumulator.
#[must_use]
pub fn new_cross_page_meta_description_accumulator() -> CrossPageMetaDescriptionAccumulator {
    CrossPageMetaDescriptionAccumulator::default()
}

/// Record a page's description. Caller passes the raw text;
/// this function trims and drops empty descriptions.
pub fn record_page_meta_description(
    acc: &mut CrossPageMetaDescriptionAccumulator,
    url: &str,
    description: &str,
) {
    let trimmed = description.trim();
    if trimmed.is_empty() {
        return;
    }
    acc.entries.push(CrossPageMetaDescriptionEntry {
        url: url.to_owned(),
        description: trimmed.to_owned(),
    });
}

/// Best-effort `origin + pathname` extraction. Sibling to the same
/// helper in `cross_page_title`; both are kept local until a third
/// aggregates detector lands and earns a shared helper extraction.
fn path_key(u: &str) -> String {
    if let Some(scheme_end) = u.find("://") {
        let after_scheme = &u[scheme_end + 3..];
        if let Some(path_start) = after_scheme.find('/') {
            let path_only = after_scheme[path_start..]
                .split_once('?')
                .map(|(p, _)| p)
                .unwrap_or(&after_scheme[path_start..])
                .split_once('#')
                .map(|(p, _)| p)
                .unwrap_or_else(|| {
                    after_scheme[path_start..]
                        .split_once('?')
                        .map(|(p, _)| p)
                        .unwrap_or(&after_scheme[path_start..])
                });
            let host = &after_scheme[..path_start];
            return format!("{}://{}{}", &u[..scheme_end], host, path_only);
        }
        return format!("{}://{}/", &u[..scheme_end], after_scheme);
    }
    u.to_owned()
}

/// Truncate description preview at 80 chars + ellipsis. Real meta
/// descriptions are commonly 150+ chars so full inclusion bloats
/// the audit detail.
fn preview(description: &str) -> String {
    let s: String = description.chars().take(80).collect();
    if description.chars().count() > 80 {
        format!("{s}…")
    } else {
        s
    }
}

/// Walk the accumulator. For each description appearing on ≥2
/// distinct URL pathnames, emit one warn finding.
#[must_use]
pub fn detect_cross_page_meta_description_duplicates(
    acc: &CrossPageMetaDescriptionAccumulator,
) -> Vec<AxisFinding> {
    let mut groups: BTreeMap<String, BTreeMap<String, String>> = BTreeMap::new();
    for entry in &acc.entries {
        let urls = groups.entry(entry.description.clone()).or_default();
        let k = path_key(&entry.url);
        urls.entry(k).or_insert_with(|| entry.url.clone());
    }

    let mut out = Vec::new();
    for (description, urls) in groups {
        if urls.len() < 2 {
            continue;
        }
        let url_list: Vec<String> = urls.into_values().collect();
        let url_preview: Vec<&str> = url_list.iter().take(5).map(String::as_str).collect();
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "meta-description.cross-page-dup".into(),
            detail: format!(
                "{} distinct URL(s) share the same <meta name=\"description\"> content '{}'. Google filters near-duplicate descriptions in search results — affected pages become effectively invisible. Social-share preview cards collapse. Set a unique, page-specific description per route. URLs: {}",
                url_list.len(),
                preview(&description),
                url_preview.join("; ")
            ),
        });
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_accumulator_no_findings() {
        let acc = new_cross_page_meta_description_accumulator();
        assert!(detect_cross_page_meta_description_duplicates(&acc).is_empty());
    }

    #[test]
    fn distinct_descriptions_no_findings() {
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "https://example.com/", "Home page");
        record_page_meta_description(&mut acc, "https://example.com/about", "About page");
        assert!(detect_cross_page_meta_description_duplicates(&acc).is_empty());
    }

    #[test]
    fn duplicate_descriptions_flag() {
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "https://example.com/", "Lorem ipsum");
        record_page_meta_description(&mut acc, "https://example.com/about", "Lorem ipsum");
        let f = detect_cross_page_meta_description_duplicates(&acc);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "meta-description.cross-page-dup");
        assert!(f[0].detail.contains("Lorem ipsum"));
    }

    #[test]
    fn query_string_variants_dedupe() {
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "https://example.com/?theme=dark", "Same page");
        record_page_meta_description(&mut acc, "https://example.com/?theme=light", "Same page");
        assert!(detect_cross_page_meta_description_duplicates(&acc).is_empty());
    }

    #[test]
    fn empty_descriptions_dropped_at_record_time() {
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "https://example.com/a", "");
        record_page_meta_description(&mut acc, "https://example.com/b", "   ");
        assert_eq!(acc.entries.len(), 0);
    }

    #[test]
    fn description_is_trimmed() {
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "https://example.com/", "  Lots of whitespace  ");
        assert_eq!(acc.entries[0].description, "Lots of whitespace");
    }

    #[test]
    fn long_description_truncated_in_preview() {
        let long = "a".repeat(200);
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "https://example.com/x", &long);
        record_page_meta_description(&mut acc, "https://example.com/y", &long);
        let f = detect_cross_page_meta_description_duplicates(&acc);
        // 80 'a's + ellipsis appears in detail; the full 200 doesn't
        assert!(f[0].detail.contains("…"));
        assert!(!f[0].detail.contains(&long));
    }

    #[test]
    fn multiple_duplicate_groups_each_get_own_finding() {
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "https://example.com/a", "Group A");
        record_page_meta_description(&mut acc, "https://example.com/b", "Group A");
        record_page_meta_description(&mut acc, "https://example.com/c", "Group B");
        record_page_meta_description(&mut acc, "https://example.com/d", "Group B");
        let f = detect_cross_page_meta_description_duplicates(&acc);
        assert_eq!(f.len(), 2);
    }

    #[test]
    fn count_reflects_distinct_urls() {
        let mut acc = new_cross_page_meta_description_accumulator();
        for p in ["a", "b", "c"] {
            record_page_meta_description(&mut acc, &format!("https://example.com/{p}"), "Same");
        }
        let f = detect_cross_page_meta_description_duplicates(&acc);
        assert!(f[0].detail.contains("3 distinct URL(s)"));
    }

    #[test]
    fn unparseable_url_falls_back_to_raw_string() {
        let mut acc = new_cross_page_meta_description_accumulator();
        record_page_meta_description(&mut acc, "raw-string-1", "Same");
        record_page_meta_description(&mut acc, "raw-string-2", "Same");
        let f = detect_cross_page_meta_description_duplicates(&acc);
        assert_eq!(f.len(), 1);
    }
}
